cdc, vector_search: enable CDC when the index is created

QuerthDP opened the pull request:
When a vector index is created in Scylla, it is initially built using a full scan of the database. After that, it stays up to date by tracking changes through CDC, which should be automatically enabled when the vector index is created.

When a user attempts to enable Vector Search (VS), the system checks whether Change Data Capture (CDC) is enabled and properly configured:

1. CDC is not enabled

    - CDC is automatically enabled with the minimum required TTL (Time-to-Live) for VS (24 hours) and the delta mode set to 'full' or post-image is enabled.

    - If the user later tries to reduce the CDC TTL below 24 hours or set delta mode to 'keys' with post-image disabled, the action fails.

    - Error message: Clearly states that CDC TTL must be at least 24 hours and delta mode must be set to 'full' or post-image must be enabled for VS to function.

2. CDC is already enabled

    - If CDC TTL is ≥ 24 hours and delta mode is set to 'full' or post-image is enabled: VS is enabled successfully.

    - If CDC TTL is < 24 hours or delta mode is set to 'keys' with post-image disabled: The VS enabling process fails.

    - Error message: Informs the user that CDC TTL must be at least 24 hours, delta mode must be set to 'full' or post-image must be enabled, and provides a link to documentation on how to update the TTL, delta mode, and post-image.

When a user attempts to disable CDC when VS is enabled, the action will fail and the user will be informed by error message that clearly states that VS needs to be disabled (vector indexes have to be dropped) first.

Full setup requirements and steps will be detailed in the documentation of Vector Search.

Co-authored-by: @smoczy123 

Fixes: VECTOR-27
Fixes: VECTOR-25

QuerthDP commented:
Changelog:

- remove VSC

TODO:

- ensure minimal set of parameters for VSC, throw an error if needed

QuerthDP commented:
Changelog:

- disallow creation of CDC and vector indexes when the Vector Search minimal CDC setup requirements are not met

QuerthDP commented on test/cqlpy/test_describe.py:
```diff
@@ -995,69 +996,88 @@ def test_table_options_quoting(cql, test_keyspace):
-        for row in ks_tables:
-            assert row.table_name == t_name or row.table_name == cdc_log_name
+
+    cdc_enablement_queries = [
```
We could've used the pytest.mark.parametrize decorator.
I will refactor the tests using it.

QuerthDP replied:
Done

QuerthDP commented on test/boost/cdc_test.cc:
```diff
@@ -210,30 +215,43 @@ SEASTAR_THREAD_TEST_CASE(test_permissions_of_cdc_log_table) {
-
-        // Disallow DROP
-        assert_unauthorized("DROP TABLE " + log_table);
+        auto cdc_enablement_queries = {
```
Do we have something like `@pytest.mark.parametrize` for boost tests? Or is it a correct approach in this case?

knowack1 commented on index/vector_index.cc:
```diff
@@ -84,6 +89,19 @@ void vector_index::validate(const schema &schema, cql3::statements::index_prop_d
+                                                               "Check documentation on how to setup delta mode - https://docs.scylladb.com/manual/branch-2025.2/features/cdc/cdc-intro.html#cdc-parameters");
+        }
+    }
+
```
I propose to refactor the function, so that each check is in dedication function:
```
check_column(targets);
check_cdc(schema);
check_options(schema);
```
This way validate() operates on a single level of abstraction.

knowack1 commented on test/cqlpy/test_cdc.py:
```diff
@@ -239,3 +239,53 @@ def test_rename_column_of_fake_cdc_log_table(cql, test_keyspace, scylla_only):
         cql.execute(f"DROP TABLE IF EXISTS {test_keyspace}.{fake_cdc_log_table_name}")
+
+
+@pytest.mark.parametrize("test_keyspace",
```
Consider to put this tests to vector search specific file instead of cdc dedicated.

knowack1 commented on test/cqlpy/test_cdc.py:
```diff
@@ -239,3 +239,53 @@ def test_rename_column_of_fake_cdc_log_table(cql, test_keyspace, scylla_only):
+@pytest.mark.parametrize("test_keyspace",
+                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
+                         indirect=True)
+def test_vs_cdc_options_ensured(scylla_only, cql, test_keyspace):
```
I prefer to have multiple small test cases/functions.

knowack1 commented on cql3/statements/describe_statement.cc:
```diff
@@ -324,7 +326,7 @@ future<std::vector<description>> table(const data_dictionary::database& db, cons
     }
 
-    if (schema->cdc_options().enabled()) {
+    if (schema->cdc_options().enabled() || secondary_index::has_vector_index(*schema)) {
```
this if `schema->cdc_options().enabled() || secondary_index::has_vector_index(*schema)` is in multiple places. I think should be  wrapped  in something like: `schema->cdc_ebabled()`. Or preferably change implementation so that `schema->cdc_options().enabled()` is sufficient check.

knowack1 commented on cdc/log.cc:
I think the cdc/log.cc file should not be aware of vector store. Instead I would try to encapsulate all the implementation in dedicated directory e.g. `scylladb/vector-search/`.
Maybe is is possible to create dedicated vector search service (scylladb/vector-search/vs_service.h`) that registers for the events to the migration notifier (as cdc service does): `_ctxt._migration_notifier.register_listener(this);` and make all checks/actions in its callbacks without touching cdc/log.cc.

QuerthDP commented on index/vector_index.cc:
```diff
@@ -84,6 +89,19 @@ void vector_index::validate(const schema &schema, cql3::statements::index_prop_d
+    auto cdc_options = schema.cdc_options();
+    if (cdc_options.enabled()) {
+        auto ttl = cdc_options.ttl();
+        if (ttl && ttl < cdc::VS_TTL_SECONDS) {
```
The messages and the case when they happen are different.
This happen when user wants to create the vector index, but CDC with non exhaustive options is enabled.
The latter check is called when user wants to create CDC but the vector index is created - it prevents the user from creating CDC not fulfilling the minimal VS setup.

knowack1 replied:
Yes, but the code is very the same and looks like a duplicate. Consider to improve it.

QuerthDP commented:
Changelog:

- moved similar code to common functions
- adjusted error messages for blocking CDC creation not meeting the VS requirements
- split the vector search CDC options test and moved to specific file
- refactored the vector index validating function to contain only one level logic

QuerthDP commented on cdc/log.cc:
```diff
@@ -176,8 +177,29 @@ class cdc::cdc_service::impl : service::migration_listener::empty_listener {
+                                                                "This means that the CDC's delta mode must be set to 'full'.\n");
+                }
+            }
+        }
```
Done

QuerthDP commented on cdc/log.cc:
```diff
@@ -205,9 +213,11 @@ class cdc::cdc_service::impl : service::migration_listener::empty_listener {
             ensure_that_table_uses_vnodes(*keyspace.metadata(), new_schema);
 
-            auto new_log_schema = create_log_schema(new_schema, log_schema ? std::make_optional(log_schema->id()) : std::nullopt, log_schema);
+            int ttl_seconds = std::max(new_schema.cdc_options().ttl(), has_vector_index ? VS_TTL_SECONDS : 0);
```
I believe it's nothing complex here that would require separate function.

QuerthDP commented:
Changelog:

- resolve valuable Copilot comments

QuerthDP commented on cdc/log.cc:
```diff
@@ -176,8 +177,29 @@ class cdc::cdc_service::impl : service::migration_listener::empty_listener {
+            }
+        }
+
+        bool is_cdc = cdc_options.enabled() || has_vector_index;
```
It is hard (if even possible) to set the `cdc_options` the way we want when creating the vector index.

My idea was that enabling vector index is just a restriction for CDC options:
- implies CDC is enabled
- implies CDC's TTL > 24h
- implies CDC's delta = full

Changing CDC options by the user is restricted or has no effect for VS (enabling, settting longer TTL, enabling pre- and postimages).
When vector index is dropped, user settings for CDC will remain unchanged, but restrictions for further option changes are down.

QuerthDP commented on test/cqlpy/test_cdc_with_vector_search.py:
I believe it's better to propagate good conventions rather than stick to the old ones.

We could have required the PEP 8 formatting for new files and keep it from now as it is a good practice to maintain readability.
@nyh what do you think? Was it discussed somewhere?

QuerthDP commented on test/cqlpy/test_cdc_with_vector_search.py:
```diff
@@ -0,0 +1,108 @@
+@pytest.mark.parametrize("test_keyspace",
+                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
+                         indirect=True)
+def test_try_create_cdc_with_vector_search_enabled(scylla_only, cql, test_keyspace):
```
As below: https://github.com/scylladb/scylladb/pull/25179#discussion_r2254123567

swasik commented:
@nyh this one is ready for review

nyh commented:
> @nyh this one is ready for review

@swasik I'll review it now.

I can't find in the cover letter or linked issues any explanation why this "24 hour or more TTL" is so critical to vector search. If a user or our support engineers believe that looking at the deployment 12 hour TTL is actually fine, why is it so important to block it?

Another thing I don't understand yet (maybe I'll understand better when I dive into the patches) is that in a previous patch which introduced a **second** CDC, it was suggested that VS needs CDC to be configured in a specific way, perhaps with post-images enabled or I'm not sure what. There is no discussion of any of that left in this patch. Can we truly use **any** CDC configuration that the user might have, and the only thing we really want to insist on is that "24 hour TTL"?

QuerthDP commented on test/cqlpy/test_vector_index.py:
```diff
@@ -10,14 +10,18 @@
 from cassandra.protocol import InvalidRequest, ConfigurationException
 
-
+@pytest.mark.parametrize("test_keyspace",
```
Changed commit message.

QuerthDP commented:
Changelog:

- resolved most of the comments
- moved tests checking correct CDC options to `test_vector_index.py`
- removed `is_vector_index` and used `dynamic_cast` to `vector_index*` instead
- changed commit message of `fix failing vector_index tests`
- moved the one time used error messages to the error throwing function instead of header file
- moved fixing test to the commit introducing changes to the logic, not the tests expansion
- renamed utility test function `create_cdc` to `alter_cdc`
- added tests showing that disabling pre- and postimage is also allowed

swasik commented:
> I agree it would be nice to have it configurable as `vector_index_minimum_cdc_ttl` with live update I suppose (?). But I'd rather have it done in a separate patch and proceed this one with what we have now.

Could you create a JIRA issue for it and link here?

QuerthDP commented:
Changelog:

- fixed error message to mention not only TTL setup, but also delta mode
- added more info about TTL and delta mode requirements to commit messages

QuerthDP commented:
> Failed Tests (109/71226):
>
>[test_ann_query_not_allow_any_filtering.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/debug___test_py___debug___test_ann_query_not_allow_any_filtering_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.debug.1)
[test_ann_query_not_allow_any_filtering.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/debug___test_py___debug___test_ann_query_not_allow_any_filtering_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.debug.2)
[test_ann_query_not_allow_any_filtering.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/debug___test_py___debug___test_ann_query_not_allow_any_filtering_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.debug.3)
[test_ann_query_not_allow_any_filtering.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/dev___test_py___dev___test_ann_query_not_allow_any_filtering_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.dev.1)
[test_ann_query_not_allow_any_filtering.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/dev___test_py___dev___test_ann_query_not_allow_any_filtering_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.dev.2)
[test_ann_query_not_allow_any_filtering.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/dev___test_py___dev___test_ann_query_not_allow_any_filtering_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.dev.3)
[test_ann_query_not_allow_any_filtering.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/release___test_py___release___test_ann_query_not_allow_any_filtering_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.release.1)
[test_ann_query_not_allow_any_filtering.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/release___test_py___release___test_ann_query_not_allow_any_filtering_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.release.2)
[test_ann_query_not_allow_any_filtering.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cqlpy/test_invalid_ann_queries/release___test_py___release___test_ann_query_not_allow_any_filtering_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ann_query_not_allow_any_filtering.release.3)
and more...

Those are the tests that were introduced here: https://github.com/scylladb/scylladb/pull/24444.
They should fail on tablets as other tests with vector index due to #16317 - since vector index creates CDC log and tablets do not yet support CDC.
I will provide a patch marking them xfail.

The 
> [test_insert_failure_doesnt_report_success.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19054/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/dev___test_py___dev___test_insert_failure_doesnt_report_success_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_insert_failure_doesnt_report_success.dev.1)

fail is a known issue: https://github.com/scylladb/scylladb/issues/25541.

QuerthDP commented:
Changelog:

- marked tests failing on tablets xfail
- allowed for using postimage enabled instead of delta set to 'full'
- rebased on master

nyh commented on index/vector_index.hh:
```diff
@@ -19,15 +19,35 @@
+    static inline const sstring NOT_SUFFICIENT_CDC_OPTIONS_TO_ENABLE_VECTOR_SEARCH =
+        format("To enable Vector Search on this table, the CDC log must meet the minimal requirements of Vector Search.\n"
+            "CDC's TTL must be at least {} seconds (24 hours), and the CDC's delta mode must be set to 'full' to enable Vector Search.\n"
+            "Check documentation on how to setup TTL - https://docs.scylladb.com/manual/branch-2025.2/features/cdc/cdc-intro.html#cdc-parameters",
```
What you're saying can be a reason not to have URLs in error messages - because if you do add URLs to error messages, you need to maintain the correctness of these links.
The problem with the approach you took is that 10 years from now, when Scylla 2035.1 comes out, we'll still have the error message pointing you to some antique document (which will likely not even exist any more). This is not even a risk - there's a certainty that this will happen (i.e., the document will either be completely outdated or not exist any more).

Personally, I think we don't need URLs in error messages at all. If the error message is clear or detailed enough it can be used in a search (or, soon, AI :-)) to find a better document on this topic. A year from now, it might find an even better document than you thought you found today.

QuerthDP commented:
> Can you please update it (and please copy-edit the first paragraph, it looks like something you stopped writing in the middle) and then I'll merge it with the correct explanation? Thanks.

Done. I think we can proceed to merge.
